feat(datasource): add Gitlab Package support

samdolt opened the pull request:
<!-- If this is your first pull request: sign the CLA with this GitHub app: https://cla-assistant.io/renovatebot/renovate -->
<!-- Make sure the `Allow edits and access to secrets by maintainers` checkbox is checked on this pull request. -->

## Changes:

<!-- Describe what behavior is changed by this PR. -->
Add a new Gitlab Package datasource

## Context:

See https://github.com/renovatebot/renovate/issues/10892 for more background

Gitlab Package API is documented here: https://docs.gitlab.com/ee/api/packages.html

This code was made by coping the `gitlab-tags` datasource, as it's doing basically the same thing, but calling another api endpoints.

It's my first time using Typescript, so surely the code could be improved



<!-- Describe why you're making these changes if it's not already explained in a corresponding issue. -->
<!-- If you're closing an existing issue with this pull request, use the keyword Closes #issue_number -->

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

Documentation would be added after the first code review

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

The code was used to create this merge request: https://gitlab.com/samdolt1/renovate-test-1/-/merge_requests/4
Unit tests are in progress

<!-- Do you have any suggestions about this PR template? Edit it here: https://github.com/renovatebot/renovate/edit/main/.github/pull_request_template.md -->

<!-- Please do not force push to your PR's branch after you have created your PR, as doing so forces us to review the whole PR again. This makes it harder for us to review your work because we don't know what has changed. -->
<!-- PRs will always be squashed by us when we merge your work. Commit as many times as you need in this branch. -->

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/renovatebot/renovate?pullRequest=10909) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/renovatebot/renovate?pullRequest=10909) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/renovatebot/renovate?pullRequest=10909) it.</sub>

samdolt commented:
One issue: Merge request generated by this datasource doesn't have a working link to the package.

Need to find a way to use the `web_path` returned by the api

```json
[
  {
    "id": 1,
    "name": "com/mycompany/my-app",
    "version": "1.0-SNAPSHOT",
    "package_type": "maven",
    "_links": {
      "web_path": "/namespace1/project1/-/packages/1",
      "delete_api_path": "/namespace1/project1/-/packages/1"
    },
```

rarkins commented:
Needs test update

viceice commented on lib/datasource/gitlab-packages/index.ts:
```diff
@@ -0,0 +1,72 @@
+  ).body;
+
+  const dependency: ReleaseResult = {
+    sourceUrl: URL.resolve(
+      depHost,
+      repo + '/-/packages?search[]=' + urlEncodedPackageName
```
This makes no sense, as source url is used for change log retrieval. Should point to the source repo of the package.

viceice commented on lib/datasource/gitlab-packages/types.ts:
```diff
@@ -0,0 +1,4 @@
+export type GitlabPackage = {
```
Prefer interface over type.

viceice commented on lib/datasource/gitlab-packages/index.ts:
```diff
@@ -0,0 +1,72 @@
+import URL from 'url';
```
Don't use deprecated url. Checkout our util/url helpers.

rarkins replied:
I'm ok to keep this *if* it's just a copy/paste from an existing datasource. We'll need to refactor those other ones anyway, and can do all the similar ones at the same time.

viceice replied:
@rarkins You told me to not make all those changes in one pr 😉 

So i would prefer to now use the right function from our utils. Should be a very small change.

viceice replied:
He already did the required change 🙃

viceice reviewed (commented):
Are you able to conver this to class based datasource, like `adoptium-java` datasource?

rarkins commented:
@viceice if it's mostly a copy/paste of an existing non-migrated datasource, are you ok to leave it as-is?

viceice commented:
> @viceice if it's mostly a copy/paste of an existing non-migrated datasource, are you ok to leave it as-is?

Sure, but i would prefer to use a already converted datasource as base, so we don't have the migration work. 🤷‍♂️

samdolt commented:
> > @viceice if it's mostly a copy/paste of an existing non-migrated datasource, are you ok to leave it as-is?
> 
> Sure, but i would prefer to use a already converted datasource as base, so we don't have the migration work. 🤷‍♂️

Maybe someone can add a note in the datasource readme.md like "For news datasource, please use 'adoptium-java' as references"
https://github.com/renovatebot/renovate/blob/main/lib/datasource/readme.md

If needed, I can probably convert both `gitlab-packages` and `gitlab-tags` to a class, because both datasource are of interest for my employer and after looking at `adoptium-java` is doesn't seem too difficult.

HonkingGoose commented on lib/datasource/gitlab-packages/readme.md:
```diff
@@ -0,0 +1,32 @@
+[Gitlab Generic Packages Repository](https://docs.gitlab.com/ee/user/packages/generic_packages/) supports looking up package versions from all type of packages registry supported by Gitlab and can be used in combination with [regex managers](https://docs.renovatebot.com/modules/manager/regex/) to keep dependencies up-to-date which are not specifically supported by Renovate.
+
+To specify which specific repository should be queried when looking up a package, the `registryUrl` has to be set like this: `https://gitlab.com/user/project`. As an example, `https://gitlab.com/gitlab-org/ci-cd/package-stage/feature-testing/new-packages-list` would look for packages in the generic packages repository of the
+`gitlab-org/ci-cd/package-stage/feature-testing/new-packages-list` project.
```
```suggestion
To specify which specific repository should be queried when looking up a package, the `registryUrl` has to be set like this: `https://gitlab.com/user/project`.
As an example, `https://gitlab.com/gitlab-org/ci-cd/package-stage/feature-testing/new-packages-list` would look for packages in the generic packages repository of the `gitlab-org/ci-cd/package-stage/feature-testing/new-packages-list` project.
```

We use the "one sentence per line" style in the docs.

HonkingGoose commented on lib/datasource/gitlab-packages/readme.md:
```diff
@@ -0,0 +1,32 @@
+
+**Usage Example**
+
+A real world example for this specific datasource would be maintaining packages versions in a config file. This can be achieved by configuring a generic regex manager in `renovate.json` for files named `versions.ini`:
```
```suggestion
A real world example for this specific datasource would be maintaining packages versions in a config file.
This can be achieved by configuring a generic regex manager in `renovate.json` for files named `versions.ini`:
```

HonkingGoose commented on lib/datasource/gitlab-packages/readme.md:
````diff
@@ -0,0 +1,32 @@
+
+```
+
+By default, the `gitlab-packages` use the `loose` versioning scheme.
````
```suggestion
By default, `gitlab-packages` uses the `loose` versioning scheme.
```

I think this is a bit better.

HonkingGoose commented:
> > > > @viceice if it's mostly a copy/paste of an existing non-migrated datasource, are you ok to leave it as-is?
> > > 
> > > 
> > > Sure, but i would prefer to use a already converted datasource as base, so we don't have the migration work. 🤷‍♂️
> > 
> > 
> > Maybe someone can add a note in the datasource readme.md like "For news datasource, please use 'adoptium-java' as references"
> > https://github.com/renovatebot/renovate/blob/main/lib/datasource/readme.md
> 
> @HonkingGoose Can you do this 🙃

See pull request #11070. 😉

viceice commented on lib/datasource/gitlab-packages/index.ts:
```diff
@@ -0,0 +1,48 @@
+export const id = 'gitlab-packages';
+export const customRegistrySupport = true;
+export const registryStrategy = 'first';
+export const defaultVersioning = 'loose';
```
Why we use `loose` versioning by default here? No other datasource is setting a default versioning like this 🤔

samdolt replied:
From Gitlab Documentation:

> package_version (String) : The package version. The following regex validates this: `\A(\.?[\w\+-]+\.?)+\z`. You can test your version strings on [Rubular](https://rubular.com/r/aNCV0wG5K14uq8).
https://docs.gitlab.com/ee/user/packages/generic_packages/index.html#publish-a-package-file

![image](https://user-images.githubusercontent.com/6305913/128150795-2a65eb7a-f5d6-44b7-a211-71b6e33cb189.png)

So theses are all valid version for a Gitlab Packages

```
word
single-word
single+word
single.1+word
v1.2.3
v1.2.3-rc4
v.1.2.3+rc5
.55.
.+.
```
So by looking at Renovate versionning documentation, it looked like the `loose` versionning was better than `semver` in this case. But maybe `docker` is better?

rarkins replied:
We have quite a few datasources where there can be non-versions in the values. GitHub tags are a simple example. But we generally take the logic of:
- Don't attempt to upgrade unless the existing value in the repo is a valid version
- Filter out all non-versions from the list of candidates

If you push ahead with `loose` I think the results will be undesirable for most. For example, if the current version is "next" then Renovate would propose to update to "stable", because it sorts later. Or even from "next" to "v1.2.3".

Better to start with docker versioning if it's likely to use similar tagging.

HonkingGoose commented on lib/datasource/gitlab-packages/index.ts:
```diff
@@ -1,48 +1,67 @@
+      ).body;
+
+      result.releases = response
+        // Settings the package_name option when calling the Gitlab API isn't enought to filter information about other package
+        // because this option is only implemented on Gitlab > 12.9 and it only do a fuzzy search.
```
```suggestion
        // Setting the package_name option when calling the GitLab API isn't enough to filter information about other package
        // because this option is only implemented on GitLab > 12.9 and it only does a fuzzy search.
```

HonkingGoose commented on lib/datasource/gitlab-packages/index.ts:
```diff
@@ -0,0 +1,78 @@
+      ).body;
+
+      result.releases = response
+        // Setting the package_name option when calling the GitLab API isn't enough to filter information about other package
```
```suggestion
        // Setting the package_name option when calling the GitLab API isn't enough to filter information about other packages
```

I think we're missing an `s` here? 😉

samdolt replied:
Yeah, I'm better with programming language than in grammer 

Thanks for the corrections :-)

HonkingGoose replied:
> Yeah, I'm better with programming language than in grammer

Funny, I'm better with grammar than programming! 😄 🙈 😜

HonkingGoose commented on lib/datasource/gitlab-packages/readme.md:
```diff
@@ -0,0 +1,33 @@
+[Gitlab Generic Packages Repository](https://docs.gitlab.com/ee/user/packages/generic_packages/) supports looking up package versions from all type of packages registry supported by Gitlab and can be used in combination with [regex managers](https://docs.renovatebot.com/modules/manager/regex/) to keep dependencies up-to-date which are not specifically supported by Renovate.
```
```suggestion
[GitLab Generic Packages Repository](https://docs.gitlab.com/ee/user/packages/generic_packages/) supports looking up package versions from all types of packages registry supported by GitLab and can be used in combination with [regex managers](https://docs.renovatebot.com/modules/manager/regex/) to keep dependencies up-to-date which are not specifically supported by Renovate.
```

I think `types` should be plural?

Capitalize nouns like GitLab, GitHub, Docker, etc. 😉

samdolt commented:
Last change was tested with `yarn test` and on a new test repository with `yarn start --platform gitlab samdolt1/renovate-test-3`

Generated MR: https://gitlab.com/samdolt1/renovate-test-3/-/merge_requests/1

HonkingGoose commented on lib/datasource/gitlab-packages/readme.md:
```diff
@@ -0,0 +1,33 @@
+
+**Usage Example**
+
+A real world example for this specific datasource would be maintaining packages versions in a config file.
```
```suggestion
A real-world example for this specific datasource would be maintaining packages versions in a config file.
```

I'm not sure which is best here:

- `real world`
- `real-world`

Can a native English speaker decide which is best? 😉

samdolt replied:
Mmh, I'm not native, but according to [Qora](https://www.quora.com/Should-real-world-be-hyphenated#:~:text=The%20simplest%20answer%20you%20can,%2C%20pronouns%20or%20noun%20phrases.%5D), in this case it's real-world.

viceice commented on lib/datasource/gitlab-packages/index.spec.ts:
```diff
@@ -0,0 +1,107 @@
+      expect(res.releases).toHaveLength(3);
+      expect(httpMock.getTrace()).toMatchSnapshot();
+    });
+    it('returns null for 404', async () => {
```
```suggestion

    it('returns null for 404', async () => {
```

readabillity

viceice commented on lib/datasource/gitlab-packages/index.spec.ts:
```diff
@@ -0,0 +1,107 @@
+      ).toBeNull();
+      expect(httpMock.getTrace()).toMatchSnapshot();
+    });
+    it('returns null for empty 200 OK', async () => {
```
```suggestion

    it('returns null for empty 200 OK', async () => {
```

viceice commented on lib/datasource/gitlab-packages/index.spec.ts:
```diff
@@ -0,0 +1,107 @@
+      ).toBeNull();
+      expect(httpMock.getTrace()).toMatchSnapshot();
+    });
+    it('throws for 5xx', async () => {
```
```suggestion

    it('throws for 5xx', async () => {
```

viceice commented on lib/datasource/gitlab-packages/index.ts:
```diff
@@ -0,0 +1,78 @@
+      this.handleGenericErrors(err);
+    }
+
+    return result.releases.length ? result : null;
```
```suggestion
    return result.releases?.length ? result : null;
```

releases is `null` by default

viceice commented on lib/datasource/gitlab-packages/index.ts:
```diff
@@ -0,0 +1,78 @@
+        .map(({ version, created_at, _links }) => ({
+          version,
+          releaseTimestamp: created_at,
+          registryUrl: _links
+            ? url.resolveBaseUrl(server, _links?.web_path)
+            : null,
```
```suggestion
```

What is this url for? If you like to addsign any package source url for change log retrival, use `sourceUrl` of `ReleaseResult` instead.

viceice commented on lib/datasource/gitlab-packages/index.ts:
```diff
@@ -0,0 +1,78 @@
+    );
+  }
+
+  async getReleases({
```
```suggestion
 @cache({
    namespace: `datasource-${datasource}`,
    key: ({ registryUrl, lookupName }: GetReleasesConfig) => `${registryUrl}-${lookupName}`,
  })
  async getReleases({
```

Add caching.

viceice commented on lib/datasource/gitlab-packages/index.ts:
```diff
@@ -0,0 +1,78 @@
+import { GitlabHttp } from '../../util/http/gitlab';
```
```suggestion
import { cache } from '../../util/cache/package/decorator';
import { GitlabHttp } from '../../util/http/gitlab';
```

See below

viceice commented on lib/datasource/gitlab-packages/readme.md:
```diff
@@ -0,0 +1,33 @@
+[GitLab Generic Packages Repository](https://docs.gitlab.com/ee/user/packages/generic_packages/) supports looking up package versions from all types of packages registry supported by GitLab and can be used in combination with [regex managers](https://docs.renovatebot.com/modules/manager/regex/) to keep dependencies up-to-date which are not specifically supported by Renovate.
```
Add a note that this requires GitLab >= 13.5

samdolt replied:
The [package api](https://docs.gitlab.com/ee/api/packages.html) was introduced in Gitlab 11.8. It's a simple API that support listing packages from all packages registries supported by Gitlab (As of now: Generic, Composer, Conan, Debian, Go Proxy, Helm, Maven, npm, nuget, PyPi and Ruby Gems. Gitlab 13.5 introduced the support for the generic package registry, but this code will also function for all other type of packages.

For performance reason, It's better to use it with Gitlab >=12.9, because of the added "package_name" argument, that allow to only get info about the package we are looking (fuzzy search by lookupName).

samdolt commented:
I will be on Vacation for the next two weeks. So if I have to make more change, it will have to wait ;-)

HonkingGoose commented on lib/datasource/gitlab-packages/readme.md:
```diff
@@ -0,0 +1,38 @@
+[GitLab Packages API](https://docs.gitlab.com/ee/api/packages.html) supports looking up package versions from [all types of packages registry supported by GitLab](https://docs.gitlab.com/ee/user/packages/package_registry/index.html) and can be used in combination with [regex managers](https://docs.renovatebot.com/modules/manager/regex/) to keep dependencies up-to-date which are not specifically s ...
+
+To specify which specific repository should be queried when looking up a package, the `registryUrl` has to be set like this: `https://gitlab.com/user/project`.
```
```suggestion
To specify which specific repository should be queried when looking up a package, the `registryUrl` has to be set following this format: `https://gitlab.com/user/project`.
```

I think we're trying to explain that you should follow a specific format when setting the `registryUrl`.

HonkingGoose commented on lib/datasource/gitlab-packages/readme.md:
```diff
@@ -0,0 +1,38 @@
+If you are using a self-hosted GitLab instance, please note the following requirements:
+
+- If you are on the `Free` edition, this datasource requires at least GitLab 13.3
+- If you are on the `Premium` or the `Ultimate` edition, this datasource requires at least GitLab 11.8, but GitLab 12.9 or more is recommended if you have a lot of packages with different name on the same project.
```
```suggestion
- If you are on the `Premium` or the `Ultimate` edition, this datasource requires at least GitLab 11.8, but GitLab 12.9 or more is recommended if you have a lot of packages with different name on the same project
```

We don't use a full stop `.` at the end of list items.

HonkingGoose commented on lib/datasource/gitlab-packages/readme.md:
```diff
@@ -0,0 +1,38 @@
+
+**Usage Example**
+
+A real-world example for this specific datasource would be maintaining packages versions in a config file.
```
```suggestion
A real-world example for this specific datasource would be maintaining package versions in a config file.
```

I think we can use the singular form here.

viceice commented on lib/datasource/gitlab-packages/index.spec.ts:
```diff
@@ -0,0 +1,107 @@
+      });
+      expect(res).toMatchSnapshot();
+      expect(res.releases).toHaveLength(3);
+      expect(httpMock.getTrace()).toMatchSnapshot();
```
```suggestion
```

We no longer need those

viceice commented on lib/datasource/gitlab-packages/index.spec.ts:
```diff
@@ -0,0 +1,107 @@
+          depName: 'mypkg',
+        })
+      ).toBeNull();
+      expect(httpMock.getTrace()).toMatchSnapshot();
```
```suggestion
```

viceice commented on lib/datasource/gitlab-packages/index.spec.ts:
```diff
@@ -0,0 +1,107 @@
+          depName: 'mypkg',
+        })
+      ).toBeNull();
+      expect(httpMock.getTrace()).toMatchSnapshot();
```
```suggestion
```

viceice commented on lib/datasource/gitlab-packages/index.spec.ts:
```diff
@@ -0,0 +1,107 @@
+          depName: 'mypkg',
+        })
+      ).rejects.toThrow(EXTERNAL_HOST_ERROR);
+      expect(httpMock.getTrace()).toMatchSnapshot();
```
```suggestion
```

viceice commented on lib/datasource/gitlab-packages/index.ts:
```diff
@@ -0,0 +1,81 @@
+export class GitlabPackagesDatasource extends Datasource {
+  static readonly id = datasource;
+
+  protected http: GitlabHttp;
```
@JamieMagee Looks like this is enough to override the base type 🙃

viceice commented:
@samdolt It seems you've disabled maintainer push acces to your fork. So we can't push our suggestions. Please enable the checkbox. So we can push changes to the pr branch.

samdolt commented:
> @samdolt It seems you've disabled maintainer push acces to your fork. So we can't push our suggestions. Please enable the checkbox. So we can push changes to the pr branch.

Mmh, I don't see that checkbox. After invistigating, this feature is only available for forks that are owned by a Github user, but I instead forked it to an organisation. (see https://github.community/t/how-can-we-enable-allow-edits-from-maintainers-by-default/2847/9). Good to know for my next pull-request

HonkingGoose commented on lib/datasource/gitlab-packages/readme.md:
```diff
@@ -0,0 +1,38 @@
+If you are using a self-hosted GitLab instance, please note the following requirements:
+
+- If you are on the `Free` edition, this datasource requires at least GitLab 13.3
+- If you are on the `Premium` or the `Ultimate` edition, this datasource requires at least GitLab 11.8, but GitLab 12.9 or more is recommended if you have a lot of packages with different name on the same project
```
```suggestion
- If you are on the `Premium` or the `Ultimate` edition, this datasource requires at least GitLab 11.8, but GitLab 12.9 or more is recommended if you have a lot of packages with different names on the same project
```

I think we should use the plural form here?

rarkins commented:
Needs de-conflicting

viceice reviewed (changes requested):
Needs some special host rule handling

See #11605, you need to update the constants and extend the test.

viceice reviewed (changes requested):
Something went wrong on merge, you need to rebase onto main and force-push

samdolt commented:
> Something went wrong on merge, you need to rebase onto main and force-push

I removed the merge commit and rebased on the latest main. Should be working now

viceice commented on lib/datasource/gitlab-packages/index.ts:
```diff
@@ -0,0 +1,81 @@
+    registryUrl: string,
+    lookupName: string
+  ): string {
+    const parsedRegistryUrl = url.parseUrl(registryUrl);
+
+    const packageName = encodeURIComponent(lookupName);
+
+    const server = parsedRegistryUrl.origin;
+    const project = encodeURIComponent(parsedRegistryUrl.pathname.substring(1)); // remove leading /
+
+    return url.resolveBaseUrl(
+      server,
+      `/api/v4/projects/${project}/packages?package_name=${packageName}&per_page=100`
+    );
```
I would prefer that the registry url is the path to gitlab only and use `lookupName` as project.

So i can host gitlab in a sub path, eg `https://domain.test/gitlab` won't currently work

viceice replied:
also use use `url.joinUrlParts` instead of `resolveBaseUrl`

viceice commented on lib/datasource/gitlab-packages/index.spec.ts:
```diff
@@ -0,0 +1,102 @@
+        })
+        .reply(200, body);
+      const res = await getPkgReleases({
+        datasource,
+        registryUrls: ['https://gitlab.com/user/project1'],
+        depName: 'mypkg',
```
```suggestion
        datasource,
        registryUrls: ['https://gitlab.com'],
        depName: 'mypkg',
        lookupName: 'user/project1'
```

See comment below

viceice commented:
You shouldn't open pr's from organization forks. We are not able to update your branch with latest main changes, so it can delay merging needlessly

rarkins commented:
> You shouldn't open pr's from organization forks. We are not able to update your branch with latest main changes, so it can delay merging needlessly

Unfortunately this is not well documented by GitHub, so it's an easy "mistake" to make.

samdolt commented:
> You shouldn't open pr's from organization forks. We are not able to update your branch with latest main changes, so it can delay merging needlessly

So I will applies your suggestion, then open a new pull request with the same content from my Github account.
